Workhop instructions (#2)
ScilifelabDataCentre/dds_workshop: docs/2_registration_project_access_session.md

# 2. Registration and project access

You should have received a registration link from 

> services-noreply@scilifelab.se

This link is to register a new account, in the testing instance and in a custom unit for this workshop. Open the link and create a new account.

Once that is done. Run
~~~
dds auth login
~~~

Examples on how to use it are available in the [documentation.](https://scilifelabdatacentre.github.io/dds_cli/examples/#authentication-dds-auth)

> ![enter image description here](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-auth-login.svg)

Reviewer: The default accessibility alternate text still there - could add a brief description of the image.

>  - [ ] TASK: Find the email and create an account, login filling your information. Note that the password will not being shown while you are typing it.

Reviewer: Grammar 

```suggestion
 - [ ] TASK: Find the email and create an account and login using your information. Note that the password will not be shown while you are typing it.
```

# Create a project

~~~
dds project create --title "<Project Title>" --description "<Project Description>" --principal-investigator "<Email to PI>"
~~~

> > The email specified in the option `--principal-investigator` does not receive any emails; It’s only for information purposes at this time.

Reviewer: 2 things: 
1) Might also be good to note that this doesn't automatically create an account for the PI (nor does it care whether a DDS a/c does exist with that address)
2) Capitalisation 
```suggestion
> The email specified in the option `--principal-investigator` does not receive any emails; it’s only for information purposes at this time.
```

[Docs.](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-create)
When the project is created, you should get an output similar to the one below. Remember the **Project ID**

> ![enter image description here](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-project-create.svg)

Reviewer: Accessible image description missing

>  - [ ] TASK: Run the create project command, remember to change the files of **title**, **description** and **pi**

Reviewer: Did we want to do anything here about the non-required options (researcher, owner and non-sensitive)?

I'm guessing there will be some people who have never used DDS but we will also have people who do already use it so might be good to try and make them aware of some options they don't already use.

Author: I think those more advanced questions, we can mention them live. Specially if they ask for them.

It will be difficult to account for all the specific options each command has, but all the commands explained have links to the docs, where all the flags are explained. And we can project the documentation while we explain the commands.

Reviewer: I think if we're going to talk through it, we should have it written down here (possibly in a box or section labelled Optional). Given that there will be ~15 people, they likely won't remain perfectly in sync with each other and there will be people who won't be able to keep up with the live presentation aspect

Author: I will mention it as a note, okay

# Upload data

By default, the project will be "In progress" (which means that data can be uploaded, but not downloaded).

Check the folder which contains the data, inside this repository.
~~~

> ls -R data/

Reviewer: This command won't work on Windows. 

Need to use `dir`

~~~

> example_directory_1     example_directory_2
> 
> data/example_directory_1: sub_directory_1
> 
> data/example_directory_1/sub_directory_1: example_file_1.txt
> 
> data/example_directory_2: example_file_2.txt      example_file_5.txt  
> sub_directory_2
> 
> data/example_directory_2/sub_directory_2: example_file_3.txt     
> example_file_4.txt

The general upload command is [`dds data put`.](https://scilifelabdatacentre.github.io/dds_cli/data/#dds-data-put)
~~~

> dds data put --project "<Project ID>" --source "<File or directory to upload>"

Reviewer: Same comment as above: it might be cool to pick 1 or 2 of the options (especially if there are some that are more useful) and have users try them out too

Author: Some of the other flags from put are way more advanced.
I prefer to keep this tutorial simple. And during the session explain the documentation and the options.

Reviewer 2: I agree with both of your points here, maybe we can have a task during the session where they look at the options and try to figure them out and/or raise questions about them

Reviewer: Same as previous comment - realistically not all the attendees will be able to keep pace (and some will already be advanced users so will be familiar with these commands). If we are going to talk about the options then I think we should add a little section on that (but possibly labelling it optional)

Author: The optional flags for this command can be tricky, I can mention as a note that there are other options they can look up in the documentation (which they already have linked) but explain them can confuse more

> dds data put --project "<Project ID>" --source "<File or directory to upload>"

Reviewer: I think we should tell them about considering zipping before upload (because we want them to do that)

~~~

 - [ ] TASK: Upload the files in the data folder. Remember to change the **project ID** and the **source** from the command above

If you have forgotten the project ID, you can list all active projects you have access with: [Docs](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-ls)

~~~
dds project ls
~~~

# List contents

> You can interactively list the contents of a project with the [`ls` command:](https://scilifelabdatacentre.github.io/dds_cli/data/#dds-data-ls)

Reviewer: What does 'interactively' mean here?

Author: That the command prompts them again to specify a subdirectory to navigate through the files

~~~
dds data ls --project "<Project ID>"
~~~

> However, You can view all files and directories in a project as a tree structure, using the `--tree` option.

Reviewer: ```suggestion
However, you can view all files and directories in a project as a tree structure, using the `--tree` option.
```

~~~
dds data ls --project "<Project ID>" --tree
~~~
![enter image description here](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-data-ls-tree.svg)

>  - [ ] TASK: Verify that all the data has being uploaded sucesfully

Reviewer: Typo 

```suggestion
 - [ ] TASK: Verify that all the data has being uploaded successfully
```

>  - [ ] TASK: Verify that all the data has being uploaded sucesfully

Reviewer: Do we have any tips for users (possibly not DDS commands but Linux or Windows CMD commands) for sanity checking the upload given that some uploads might have 10,000s of files? 
For example, commands to count the number of files in a dir (e.g. `ls . | wc -l`) or get the total size of all files in the directory (e.g. `du -hs ./*`)

Author: Actually, you reminded me of something. 
One of these commands shows the size as Kibibytes, and our system stores the data as kilobytes (or the other way around, don't remember exactly). So when big uploads, there is a big difference on size if someone manually checks like that. Need to check it

# Release project and download data

> 1. In order for researchers to get access to the data, you need to [change the status](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-access) to `Available`.

Reviewer: Might be interesting to show them (even if we don't get them to test it) what researchers see if they try to download data when the project is not set to Available.

Author: It's not so mysterious, the client just stops and tells them to wait until the project is available (has happened a lot of times to me while locally testing).

Also, they get an email when it is released. So they know about it.

Author: I added a small explanation as a note

> 1. In order for researchers to get access to the data, you need to [change the status](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-access) to `Available`.

Reviewer: Might also be good to mention that putting the project into Available starts a countdown in number of days

Author: Should be mentioned in project life cycle

~~~
dds project status release --project "<Project ID>"
~~~

![enter image description here](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-project-status-release.svg)
##

2. To [invite researchers](dds%20user%20add%20%5BEmail%20address%5D%20--role "&lt;Account role&gt;quot; --project quot;&lt;Project ID&gt;") that will download the data to the project:

~~~
dds project access grant --project "<Project ID>" --email "<email>"
~~~

> or

Reviewer: Might be good to note what the difference is: when should they pick one over the other?

Author: No difference at all, just two commands that do the same. The second one can be used to add any other use if they change the role. But for this purposes it is the same.

Reviewer 2: Isn't the first command intended for existing accounts, i.e. it should fail if there is no account in the system with the specified email address (not sure if that's the behaviour, we should check it)

Author: Clarified that they do the same along with a reminder that they can also invite researcher when creating project

Reviewer 3: So I checked the code a bit quickly now and the `dds project access grant` command calls the `AddUser` resource / endpoint, which is the same one that `dds user add` calls, so there's no difference in functionality. I guess the idea is that they use `dds user add` (which was the only command from the beginning) as a default when they're inviting new accounts and that they use `dds project access grant` when they want to give an existing account access, but they could technically use both for both purposes. Not sure atm what's best but I feel like I personally would just want to know about one command first ("the intended way") and then possibly have a note somewhere where it says 'you could also use xxx', maybe including both here can cause some confusion / more questions.

Reviewer 2: I believed that `dds project access grant --project "<Project ID>" --email "<email>"` fail if the specified email doesn't already have an account registered, but it seems I was wrong. Just checked it and it turns out it sends an invitation if there's no account. 
We can leave it as it is, with minor change in the text

~~~
dds user add [Email address] --role "Researcher" --project "<Project ID>"
~~~

##
3. To download the full project contents
~~~
dds data get --get-all --project "<Project ID>" 
~~~

You can also use the `--source` option to specify which file or directory you want to download within the project. If you want to download multiple individual files or directories, specify the `--source` option multiple times.

~~~
dds data get --source "<1st file or directory>" --source "<2nd file or directory>" [... etc] --project "<Project ID>"
~~~

 - [ ] TASK: Release the project after making sure all the data was uploaded, remember to specify the correct project ID
 - [ ] TASK: Download the project contents, and verify that they are the correct. You can download data with your unit admin account.

>  - [ ] Optional: Use another email you may have to invite yourself as a researcher, you don't need to register again.

Reviewer: I get what you mean here but for some reason find this sentence hard to parse. Maybe could say:
_If you have another account on Test DDS, you can invite that email as a researcher_
That said, 1) couldn't they just use plus addressing to instantly invite themselves as a researcher anyway? 2) how does it work when you want to set up authentication for 2 different accounts on the same computer? Presumably they'd have to end/delete the stored session token and then re-auth with the other account's credentials. Might be good to mention that

Author: It is just so they could test adding emails as a researcher, I don't intend on them to log in again. Because that could be too much mess.

Reviewer: I still don't really understand this one.

Reviewer 2: Yes, this is one of the places where the flow gets confusing to me. Maybe we should rearrange things, because the section title is "Release project and download data" but it includes giving access to researchers. Probably cleaner if we keep things to unit personnel here, i.e. release project and download data

Reviewer 3: Let me know if I'm misinterpreting etc.

But basically this is how it works:
- If they want to be able to download with a researcher account then they'd need to invite themselves as a researcher to that project, and then that researcher email would need to register an account. If that email has already been invited to the system and they have registered a researcher account previously, then they can invite that email to a project, and they then automatically have access to download, they wouldn't need to register again.
- In order to download with that researcher account, first they'd need to re-auth with the researcher credentials, and that would replace the existing local token (that belonged to the authenticated unit user account). 

So I think the "you don't need to register again" should be removed/moved somewhere else? If we want to have this optional suggestion here we could rephrase it to something like this, perhaps? (But we may want to include steps in that case)
> - [ ] Optional: If you have another email address, use your Unit Admin / Personnel account to invite that email as a researcher. Once you have registered the "researcher" account, authenticate using the researcher account credentials and use the `dds data get` command as described above to download the data.

Reviewer 2: We should either change the text as Ina suggests, or don't mess things with Researcher account here. 
Maybe the best is to make this change now, and later on decide if it should go to another section or removed all together.

![Screenshot of a sucessfull project creation](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-project-create.svg)

 - [ ] TASK: Run the create project command, remember to change the files of **title**, **description** and **pi**

# Upload data

By default, the project will be "In progress" (which means that data can be uploaded, but not downloaded).

Check the folder which contains the data, on Mac/Linux, inside this repository.
~~~
ls -R data/
~~~

> example_directory_1     example_directory_2
> 
> data/example_directory_1: sub_directory_1
> 
> data/example_directory_1/sub_directory_1: example_file_1.txt
> 
> data/example_directory_2: example_file_2.txt      example_file_5.txt  
> sub_directory_2
> 
> data/example_directory_2/sub_directory_2: example_file_3.txt     
> example_file_4.txt

On windows, you can use the `dir` command or manually explore the structure through the file explored.

The general upload command is [`dds data put`.](https://scilifelabdatacentre.github.io/dds_cli/data/#dds-data-put)
~~~
dds data put --project "<Project ID>" --source "<File or directory to upload>"
~~~

> > NOTE: As of today, DDS can break when there are heavy uploads. We are working on fixing it, but on the meantime, we recommned to ZIP files when the uploads are big (Hundreds of files).

Reviewer 2: ```suggestion
> NOTE: As of today, DDS has performance issues and upload failures when uploading many small files. We are working on fixing it, but on the meantime, we recommend you to archive (with ZIP or RAR, no compression required) files when the delivered data consists of hundreds or thousands of small files).
```

![Screenshot of the files structure in the project](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-data-ls-tree.svg)

 - [ ] TASK: Verify that all the data has being uploaded succesfully.

# Release project and download data

> Currently DDS does not support to resume downloads, if your download is interrupted you will need to restart it.

1. In order for researchers to get access to the data, you need to [change the status](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-access) to `Available`. If a researcher tries to download data and the project is stil in progress they will get an error.

~~~
dds project status release --project "<Project ID>"
~~~

![enter image description here](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-project-status-release.svg)
##

After releasing the project, any curent researcher associated with the project will recieve a notification by email.

2. To [invite researchers](dds%20user%20add%20%5BEmail%20address%5D%20--role "&lt;Account role&gt;quot; --project quot;&lt;Project ID&gt;") that will download the data to the project:

~~~
dds project access grant --project "<Project ID>" --email "<email>"
~~~

> or

Reviewer 2: ```suggestion
for giving access to an already existing account, or
```

> or

Reviewer: @valyo - given this suggestion, would you disagree with the text (see discussion immediately above this one) that says the commands are the same? 
Edit: I can see from below that yes you disagree there is no difference.

Reviewer 2: The end result is the same in terms that a researcher account gets access to a project, but the commands are for different scenarios. So I wouldn't say the commands are the same. Maybe we don't even need to say explicitly that they are the same or different. I will thy to find a solution

~~~
dds user add [Email address] --role "Researcher" --project "<Project ID>"
~~~

> 

Reviewer 2: ```suggestion

for sending an invitation for a new account, which will automatically obtain access to that specific project.

```

> >There is no practical difference between this two command in this use case. However, as mentioned above, it is possible to automatically invite researchers when creating  a project. \

Reviewer 2: ```suggestion
>As mentioned above, it is possible to automatically invite researchers when creating  a project. \
```

# Upload data

> When an upload is interrupted, you can resume it later on. The system will detect which files are already on the cloud and only ulpload the remaining ones. In order to replace them you must specify the `--overwrite` flag.

By default, the project will be "In progress" (which means that data can be uploaded, but not downloaded).

Check the folder which contains the data, on Mac/Linux, inside this repository.
~~~
ls -R data/
~~~

> example_directory_1     example_directory_2
> 
> data/example_directory_1: sub_directory_1
> 
> data/example_directory_1/sub_directory_1: example_file_1.txt
> 
> data/example_directory_2: example_file_2.txt      example_file_5.txt  
> sub_directory_2
> 
> data/example_directory_2/sub_directory_2: example_file_3.txt     
> example_file_4.txt

On windows, you can use the `dir` command or manually explore the structure through the file explored.

The general upload command is [`dds data put`.](https://scilifelabdatacentre.github.io/dds_cli/data/#dds-data-put)
~~~
dds data put --project "<Project ID>" --source "<File or directory to upload>"
~~~

> > There are a number of other optional flags, which cab be found on the documentation link just above.

Reviewer 2: ```suggestion
> There are a number of other optional flags, which can be found on the documentation link just above.
```

# Release project and download data

> Currently DDS does not support to resume downloads, if your download is interrupted you will need to restart it.

1. In order for researchers to get access to the data, you need to [change the status](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-access) to `Available`. If a researcher tries to download data and the project is stil in progress they will get an error.

~~~
dds project status release --project "<Project ID>"
~~~

![Message shown after project release](https://scilifelabdatacentre.github.io/dds_cli/_images/dds-project-status-release.svg)
##

After releasing the project, any curent researcher associated with the project will recieve a notification by email.

2. To [invite researchers](https://scilifelabdatacentre.github.io/dds_cli/project/#dds-project-access-grant) that will download the data to the project:

~~~
dds project access grant --project "<Project ID>" --email "<email>"
~~~

or

~~~
dds user add [Email address] --role "Researcher" --project "<Project ID>"
~~~

> There is no practical difference between these two commands for this use case. However, as mentioned above, it is 

> > possible to automatically invite researchers when creating a project. `dds project create (...) --researcher [Email address]`

Reviewer 2: ```suggestion
> There is no practical difference between these two commands for this use case. As mentioned above, it is 
> also possible to automatically invite researchers when creating a project. `dds project create (...) --researcher [Email address]`
```

##
3. To [download](https://scilifelabdatacentre.github.io/dds_cli/data/#dds-data-get) the full project contents
~~~
dds data get --get-all --project "<Project ID>" 
~~~

You can also use the `--source` option to specify which file or directory you want to download within the project. If you want to download multiple individual files or directories, specify the `--source` option multiple times.

~~~
dds data get --source "<1st file or directory>" --source "<2nd file or directory>" [... etc] --project "<Project ID>"

> ~~~

Reviewer 2: This one should be in the next section, IMO. 
Since it is too messy to suggest the change here, I will do it locally and commit it

 - [ ] TASK: Release the project after making sure all the data was uploaded, remember to specify the correct project ID
 - [ ] TASK: Download the project contents, and verify that they are the correct. You can download data with your unit admin account.

>  - [ ] Optional: Use another email you may have to invite yourself as a researcher, you don't need to register again.

Reviewer 2: ```suggestion
 - [ ] Optional: If you have another email address, use your Unit Admin / Personnel account to invite that email as a researcher. Once you have registered the "researcher" account, authenticate using the researcher account credentials and use the dds data get command as described above to download the data.
```
